precompiles/dapp-staking-v3/src/lib.rs

```diff
@@ -0,0 +1,832 @@
+        )?;
+
+        let smart_contract =
+            <R as pallet_dapp_staking_v3::Config>::SmartContract::evm(contract_h160.into());
+
+        // parse the staker account
+        let staker = Self::parse_input_address(staker.into())?;
+
+        // Get staking info for the staker/contract combination
+        let staking_info = StakerInfo::<R>::get(&staker, &smart_contract).unwrap_or_default();
+        log::trace!(target: "ds-precompile", "read_staked_amount_on_contract for account:{:?}, staking_info: {:?}", staker, staking_info);
+
+        // Ensure that the staking info is checked against the current period (stakes from past periods are reset)
+        let current_period_number = ActiveProtocolState::<R>::get().period_number();
+
+        if staking_info.period_number() == current_period_number {
+            Ok(staking_info.total_staked_amount())
+        } else {
+            Ok(0_u128)
+        }
+    }
+
+    /// Read the total amount staked on the given contract right now.
+    #[precompile::public("read_contract_stake(address)")]
+    #[precompile::view]
+    fn read_contract_stake(
+        handle: &mut impl PrecompileHandle,
+        contract_h160: Address,
+    ) -> EvmResult<u128> {
+        // TODO: benchmark this function so we can measure ref time & PoV correctly
+        // Storage item: ActiveProtocolState:
+        // Twox64(8) + ProtocolState::max_encoded_len
+        // Storage item: IntegratedDApps:
+        // Blake2_128Concat(16 + SmartContract::max_encoded_len) + DAppInfoFor::max_encoded_len
+        // Storage item: ContractStake:
+        // Twox64Concat(8) + EraIndex(4) + ContractStakeAmount::max_encoded_len
+        handle.record_db_read::<R>(
+            36 + ProtocolState::max_encoded_len()
+                + <R as pallet_dapp_staking_v3::Config>::SmartContract::max_encoded_len()
+                + DAppInfoFor::<R>::max_encoded_len()
+                + ContractStakeAmount::max_encoded_len(),
+        )?;
+
+        let smart_contract =
+            <R as pallet_dapp_staking_v3::Config>::SmartContract::evm(contract_h160.into());
+
+        let current_period_number = ActiveProtocolState::<R>::get().period_number();
+        let dapp_info = match IntegratedDApps::<R>::get(&smart_contract) {
+            Some(dapp_info) => dapp_info,
+            None => {
+                // If the contract is not registered, return 0 to keep the legacy behavior.
+                return Ok(0_u128);
+            }
+        };
+
+        // call pallet-dapps-staking
+        let contract_stake = ContractStake::<R>::get(&dapp_info.id);
+
+        Ok(contract_stake.total_staked_amount(current_period_number))
+    }
+
+    /// Register contract with the dapp-staking pallet
+    /// Register is root origin only. This should always fail when called via evm precompile.
+    #[precompile::public("register(address)")]
+    fn register(_: &mut impl PrecompileHandle, _address: Address) -> EvmResult<bool> {
```

Comment:
Why is some exposed functions that will always fail ?

Reply:
It was part of the 'old' functionality, similar as setting the reward destination.
Not it's no longer valid, it's impossible for me to port it to v3.

So failing with a custom error seems better to me than just throwing unsupported address (not sure what the exact error would be).